src/NServiceBus.Core/Extensibility/ContextBag.cs

```diff
@@ -139,12 +139,28 @@ public void Set<T>(string key, T t)
         /// <param name="context">The source context.</param>
         internal void Merge(ContextBag context)
         {
+            if (context.stash == null)
+            {
+                return;
+            }
+
+            var targetStash = GetOrCreateStash();
             foreach (var kvp in context.stash)
             {
-                stash[kvp.Key] = kvp.Value;
+                targetStash[kvp.Key] = kvp.Value;
             }
         }
 
+        Dictionary<string, object> GetOrCreateStash()
```

Comment:
As an alternative, this could also be moved to property that initializes the field

Reply:
but then any read/delete attempt would initialize it, wouldn't it?

Reply:
The downside is that you have to sometimes use the property and sometimes use the field yes. That's why probably the current approach is "cleaner". I just wanted to mention it.

Reply:
ah, got it 👍  Yeah I think the method might be easier to understand in regards to the "side-effects"